fobnail-attester.c

```diff
@@ -88,6 +93,216 @@ coap_context_t* att_coap_new_context(const bool enable_coap_block_mode)
+                             TPMA_OBJECT_SIGN_ENCRYPT |
+                             TPMA_OBJECT_DECRYPT |
+                             TPMA_OBJECT_FIXEDTPM |
+                             TPMA_OBJECT_FIXEDPARENT |
+                             TPMA_OBJECT_SENSITIVEDATAORIGIN |
+                             TPMA_OBJECT_NODA),
+		.authPolicy.size = 0,
+		.parameters.rsaDetail = {
+			.symmetric = {
+				.algorithm = TPM2_ALG_NULL,
+				.keyBits.aes = 0,
+				.mode.aes = 0,
+			},
+			.scheme = {
+				.scheme = TPM2_ALG_NULL,
+				.details = {}
+			},
+			.keyBits = 2048,             /* Hardcoded, after many errors */
+			.exponent = 0x10001,         /* the same... */
+		},
+		.unique.rsa.size = 0
+	}
+};
+
+static TSS2_RC att_gen_id_key(void)
+{
+	TSS2_RC 	   tss_ret = TSS2_RC_SUCCESS;
+	ESYS_CONTEXT	   *esys_ctx = NULL;
+	TSS2_TCTI_CONTEXT  *tcti_ctx = NULL;
+	ESYS_TR 	   parent = ESYS_TR_NONE;
+	TPM2B_PUBLIC 	   *keyPublic = NULL;
+	TPM2B_PRIVATE	   *keyPrivate = NULL;
+	TPM2B_PUBLIC	   inPublic = keyTemplate;
+	TPM2B_SENSITIVE_CREATE inSensitive = {
+		.sensitive = {
+				.userAuth = {
+					.size = 0,
+				},
+		.data = {
+			.size = 0,
+			}
+		}
+	};
+
+	TPM2B_DATA outsideInfo = { .size = 0, };
+	TPML_PCR_SELECTION creationPCR = { .count = 0, };
+	TPM2B_DIGEST ownerauth = { .size = 0 };
+	TPMS_CAPABILITY_DATA *capabilityData = NULL;
+
+	/* Initialize the Esys context */
+	if ((tss_ret = Tss2_TctiLdr_Initialize(NULL, &tcti_ctx)) != TSS2_RC_SUCCESS) {
+		fprintf(stderr, "Error: Tss2_TctiLdr_Initialize\n");
+		goto error;
+	}
+
+	if ((tss_ret = Esys_Initialize(&esys_ctx, tcti_ctx, NULL)) != TSS2_RC_SUCCESS) {
+		fprintf(stderr, "Error: Esys_Initialize()\n");
+		goto error;
+	}
+
+	if ((tss_ret = Esys_TR_SetAuth(esys_ctx, ESYS_TR_RH_OWNER, &ownerauth)) != TSS2_RC_SUCCESS) {
+		fprintf(stderr, "Error: Esys_TR_SetAuth()\n");
+		goto error;
+	}
+
+	tss_ret = Esys_GetCapability(esys_ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+					TPM2_CAP_ALGS, 0, TPM2_MAX_CAP_ALGS, NULL, &capabilityData);
+	if (tss_ret != TSS2_RC_SUCCESS) {
+		fprintf(stderr, "Error: Esys_GetCapability()\n");
+		goto error;
+	}
+
+
+	tss_ret = Esys_CreatePrimary(esys_ctx, ESYS_TR_RH_OWNER, ESYS_TR_PASSWORD,
+				     ESYS_TR_NONE, ESYS_TR_NONE, &primarySensitive,
+				     &primaryRsaTemplate, &outsideInfo, &creationPCR,
+				     &parent, NULL, NULL, NULL, NULL);
+
+	if (tss_ret != TSS2_RC_SUCCESS) {
+		fprintf(stderr, "Error: Esys_CreatePrimary()\n");
+		goto error;
+	}
+
+
+	tss_ret = Esys_Create(esys_ctx, parent, ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
+			      &inSensitive, &inPublic, &outsideInfo, &creationPCR,
+			      &keyPrivate, &keyPublic, NULL, NULL, NULL);
+	if (tss_ret != TSS2_RC_SUCCESS) {
+		fprintf(stderr, "Error: Esys_Create()\n");
+		goto error;
+	}
+
+error:
+	if (esys_ctx != NULL)
+		Esys_Finalize(&esys_ctx);
+
+	if (tcti_ctx != NULL)
+		Tss2_TctiLdr_Finalize(&tcti_ctx);
+
+	return tss_ret;
+}
+
+
```

Comment:
Please split code across multiple source files. Packing everything into file will quickly make code unreadable.

Reply:
Fixed in new PR